repeat_expansion_variants/pipeline.py

```diff
@@ -0,0 +1,209 @@
+        # Get all identifiers we want to query BioMart with
+        identifiers_to_query = sorted({
+            i for i in variants[column_name_in_dataframe]
+            if filtering_function(i)
+        })
+        # Query BioMart for Ensembl Gene IDs
+        annotation_info = biomart.query_biomart(
+            key_column=(column_name_in_biomart, column_name_in_dataframe),
+            query_column=('ensembl_gene_id', 'EnsemblGeneID'),
+            identifier_list=identifiers_to_query,
+        )
+        # Make note where the annotations came from
+        annotation_info['GeneAnnotationSource'] = column_name_in_dataframe
+        # Combine the information we received with the *original* dataframe (a copy made before any iterations of this
+        # cycle were allowed to run). This is similar to SQL merge.
+        annotation_df = pd.merge(variants_original, annotation_info, on=column_name_in_dataframe, how='left')
+        # Update main dataframe with the new values. This replaces the NaN values in the dataframe with the ones
+        # available in another dataframe we just created, `annotation_df`.
+        variants = variants \
+            .set_index([column_name_in_dataframe]) \
+            .combine_first(annotation_df.set_index([column_name_in_dataframe]))
+
+    # Reset index to default
+    variants.reset_index(inplace=True)
+    # Some records are being annotated to multiple Ensembl genes. For example, HGNC:10560 is being resolved to
+    # ENSG00000285258 and ENSG00000163635. We need to explode dataframe by that column.
+    variants = variants.explode('EnsemblGeneID')
+    # Fetch Ensembl gene name based on Ensembl gene ID
+    annotation_info = biomart.query_biomart(
+        key_column=('ensembl_gene_id', 'EnsemblGeneID'),
+        query_column=('external_gene_name', 'EnsemblGeneName'),
+        identifier_list=sorted({str(i) for i in variants['EnsemblGeneID'] if str(i).startswith('ENSG')}),
+    )
+    variants = pd.merge(variants, annotation_info, on='EnsemblGeneID', how='left')
+    # Check that there are no multiple gene name mappings for any given gene ID
+    assert variants['EnsemblGeneName'].str.len().dropna().max() == 1, 'Multiple gene ID → gene name mappings found!'
+    # Convert the one-item list into a plain column
+    return variants.explode('EnsemblGeneName')
+
+
+def determine_repeat_type(row):
+    """
+    Based on all available information about a variant, determine its type. The resulting type can be:
+        * trinucleotide_repeat_expansion, corresponding to SO:0002165
+        * short_tandem_repeat_expansion, corresponding to SO:0002162
+        * NaN (not able to determine)
+    Also, depending on the information, determine whether the record is complete, i.e., whether it has all necessary
+    fields to be output for the final "consequences" table.
+    """
+    repeat_type = np.nan
+    if row['IsProteinHGVS']:
```

Comment:
Have you ensured that Pandas infers this column (which could have a mix of values None and True based on **clivar_identifier_parsing.parse_variant_identifier**) as a boolean?

Reply:
Yes, this issue did arise in the process of development, but in this PR it is already fixed. In `parse_variant_identifier`, is_protein_hgvs is always assigned either False or True, never None:
https://github.com/tskir/vep-mapping-pipeline/blob/74339dfa546dfce2a3fa5b355f02f73981895b0b/repeat_expansion_variants/clinvar_identifier_parsing.py#L87 (the assignments go None, None, None, **False**)